### DISCUSSION TITLE
Mixed boundary conditions

### DISCUSSION THREAD

**Discussion by WenjieYao** [kinnala/scikit-fem#1002] [Q&A]
I want to solve a problem with periodic boundary conditions on the left and right side, the top and bottom side have nonzero Dirichlet boundary conditions in some segments and Neumann boundary conditions in other segements, how should I do that? I tried some codes from expample 13 and 42, but I got errors `Transforming over 1000 vertices to C_CONTIGUOUS.` when I execute `m = MeshTri1DG.init_tensor(
    np.linspace(0, 1, 30),
    np.linspace(0, 1, 30),
    periodic=[0],
).with_boundaries({
            'ground': lambda x: x[1] == 0.,
            'positive': lambda x: x[1] == 1,
        })`

**Comment by kinnala** [comment]
`Transforming over 1000 vertices to C_CONTIGUOUS` is not an error but a warning. It is of no concern if you do not know what it means. I'll transform this to Discussions.

**Comment by kinnala** [comment]
I suggest you don't use `MeshTri1DG` because there might be unexpected results if your segments are touching the periodic boundary.

Try to instead use the following snippet as a starting point:
```python
import numpy as np
from skfem import *
from skfem.models import laplace, unit_load

m = MeshTri().refined(4)
e = ElementTriP1()

basis = Basis(m, e)

left = basis.get_dofs(facets='left').nodal['u']
right = basis.get_dofs(facets='right').nodal['u']
D = basis.get_dofs({'top', 'bottom'}).all()

left = np.setdiff1d(left, D)
right = np.setdiff1d(right, D)

B = np.zeros((len(left), basis.N))
B[range(len(left)), left] = 1
B[range(len(right)), right] = -1

A = asm(laplace, basis)
f = asm(unit_load, basis)

# create a saddle point system with one additional Lagrange multiplier / periodic DOF pair
A = bmat([[A, B.T],
          [B, None]], 'csr')
f = np.concatenate((f, 0 * left))

y = solve(*condense(A, f, D=basis.get_dofs({'top', 'bottom'})))

basis.plot(y[:basis.N], colorbar=True, shading='gouraud', levels=5).show()
```
It uses Lagrange multipliers to enforce periodicity.

**Reply by WenjieYao** [reply]
I also tried to use `MeshTri.load` to load periodic mesh created from gmsh, but it seems this function is not supported yet, it only loads a ordinary square domain even if there is `$Periodic` information in the `.msh` file.  I think it would be a very important feature to add that recoganize a periodic mesh, which can simplify things a lot

**Reply by kinnala** [reply]
We do reading of meshes using another package called meshio. I'm quite sure periodic meshes are not supported by meshio.

**Reply by kinnala** [reply]
Perhaps meshio can read periodicity information into meshio cell_data or point_data. If that's the case then we are able to use the information.

**Reply by kinnala** [reply]
Can you upload a simple example of periodic .msh so we can try it out?

**Reply by WenjieYao** [reply]
I am using the following code to generate the mesh:
```
import gmsh
from dataclasses import dataclass


@dataclass
class PeriodicGeometry:
    L: float = 10
    H: float = 10
    lc: float = 2


def mesh_generator(g: PeriodicGeometry, file_name: str):
    gmsh.initialize()
    gmsh.option.setNumber("General.Terminal", 1)
    gmsh.option.setNumber("Mesh.Algorithm", 6)
    gmsh.clear()
    gmsh.model.add("geometry")

    # Add points
    p1 = gmsh.model.geo.addPoint(-g.L / 2, -g.H / 2, 0, g.lc / 2)
    p2 = gmsh.model.geo.addPoint(+g.L / 2, -g.H / 2, 0, g.lc / 2)
    p3 = gmsh.model.geo.addPoint(+g.L / 2, +g.H / 2, 0, g.lc / 2)
    p4 = gmsh.model.geo.addPoint(-g.L / 2, +g.H / 2, 0, g.lc / 2)

    # Add lines
    l1 = gmsh.model.geo.addLine(p1, p2)
    l2 = gmsh.model.geo.addLine(p2, p3)
    l3 = gmsh.model.geo.addLine(p3, p4)
    l4 = gmsh.model.geo.addLine(p1, p4)
    # Construct curve loops and surfaces
    c1 = gmsh.model.geo.addCurveLoop([l1, l2, l3, -l4])
    s1 = gmsh.model.geo.addPlaneSurface([c1])
    gmsh.model.geo.synchronize()
    # Physical groups
    gmsh.model.addPhysicalGroup(1, [l1], name="bottom")
    gmsh.model.addPhysicalGroup(1, [l2], name="right")
    gmsh.model.addPhysicalGroup(1, [l3], name="top")
    gmsh.model.addPhysicalGroup(1, [l4], name="left")
    gmsh.model.addPhysicalGroup(2, [s1], name="all")

    # Set periodic mesh on the left and right side
    gmsh.model.mesh.setPeriodic(1, [l2], [l4],
                                [1, 0, 0, g.L, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1])
    # We can then generate a 2D mesh...
    gmsh.model.mesh.generate(2)

    # ... and save it to disk
    gmsh.write(file_name)
    gmsh.finalize()


if __name__ == '__main__':
    g = PeriodicGeometry()
    mesh_generator(g, 'periodic.msh')

```

**Reply by kinnala** [reply]
Here is the file: [periodic.msh.txt](https://github.com/kinnala/scikit-fem/files/10360132/periodic.msh.txt)

**Reply by kinnala** [reply]
This is how to load it using `meshio`:
```python
In [15]: import meshio

In [16]: meshio.read("periodic.msh")

Out[16]: 
<meshio mesh object>
  Number of points: 142
  Number of cells:
    line: 10
    line: 10
    line: 10
    line: 10
    triangle: 242
  Cell sets: bottom, right, top, left, all, gmsh:bounding_entities
  Point data: gmsh:dim_tags
  Cell data: gmsh:physical, gmsh:geometrical
  Field data: bottom, right, top, left, all
```
I cannot find here anything which refers to it being a periodic mesh. Perhaps one of those datasets with `gmsh:` prefix has the information.

**Reply by kinnala** [reply]
Anyways, since you have those `left` and `right` tags, it should be easy to use those to solve a periodic problem. Just copypaste the above code I gave you:
```python
import numpy as np
from skfem import *
from skfem.models import laplace, unit_load

m = MeshTri.load("periodic.msh")
e = ElementTriP1()

basis = Basis(m, e)

left = basis.get_dofs(facets='left').nodal['u']
right = basis.get_dofs(facets='right').nodal['u']
D = basis.get_dofs({'top', 'bottom'}).all()

left = np.setdiff1d(left, D)
right = np.setdiff1d(right, D)

B = np.zeros((len(left), basis.N))
B[range(len(left)), left] = 1
B[range(len(right)), right] = -1

A = asm(laplace, basis)
f = asm(unit_load, basis)

# create a saddle point system with one additional Lagrange multiplier / periodic DOF pair
A = bmat([[A, B.T],
          [B, None]], 'csr')
f = np.concatenate((f, 0 * left))

y = solve(*condense(A, f, D=basis.get_dofs({'top', 'bottom'})))

basis.plot(y[:basis.N], colorbar=True, shading='gouraud', levels=5).show()
```

**Reply by WenjieYao** [reply]
I think mehsio doesn't read the periodic information from a `gmsh` created file, you need to import `gmsh` to get the information...

**Reply by kinnala** [reply]
Ok, then this is a feature request for `meshio` since we won't import `gmsh` in `scikit-fem` due to its license.
